File: Source/Events/Gameover.cs
using System;
using WCSharp.Api;

public static class Gameover
{
    public static bool WinGame { get; set; } = false;
    private static float EndingTimer { get; set; } = 90.0f;
    public static bool NoEnd { get; set; } = false;

    public static bool GameOver()
    {
        return WinningGame() || LosingGameCheck();
    }

    private static bool WinningGame()
    {
        if (!WinGame) return false;
        SendWinMessage();
        GameStats(true);
        GameoverUtil.SetColorData();
        GameoverUtil.SetBestGameStats();
        GameoverUtil.SetFriendData();
        StandardWinChallenges();
        SaveGame();
        Console.WriteLine($"{Colors.COLOR_GREEN}Stay a while for the end game awards!!{Colors.COLOR_RESET}");
        Utility.SimpleTimer(5.0f, PodiumManager.BeginPodiumEvents);
        return true;
    }

    private static void StandardWinChallenges()
    {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return;
        Challenges.NecroWindwalk();
        Challenges.BlueFire();
        Challenges.PinkFire();
        Challenges.WhiteTendrils();
        Challenges.ZandalariKitty();
        Challenges.FreezeAura();
    }

    private static void LosingGame()
    {
        Wolf.RemoveAllWolves();
        GameoverUtil.SetColorData();
        GameoverUtil.SetFriendData();
        GameStats(false);
        SaveGame();
        NotifyEndingGame();
    }

    private static void SaveGame()
    {
        Utility.SimpleTimer(1.5f, SaveManager.SaveAll);
        Utility.SimpleTimer(2.5f, SaveManager.SaveAllDataToFile);
    }

    private static void EndGame()
    {
        foreach (var player in Globals.ALL_PLAYERS)
            Blizzard.CustomVictoryBJ(player, true, true);
    }

    private static bool LosingGameCheck()
    {
        if (Gamemode.CurrentGameMode != GameMode.Standard) return false;
        if (NoEnd) return false;

        for(int i = 0; i < Globals.ALL_PLAYERS.Count; i++)
        {
            var kitty = Globals.ALL_KITTIES[Globals.ALL_PLAYERS[i]];
            if (kitty.Alive) return false;
        }
        LosingGame();
        return true;
    }

    private static void SendWinMessage()
    {
        if (Gamemode.CurrentGameMode == GameMode.Standard)
            Console.WriteLine($"{Colors.COLOR_GREEN}Congratulations on winning the game on {Difficulty.DifficultyOption.ToString()}!{Colors.COLOR_RESET}");
        else
            Console.WriteLine($"{Colors.COLOR_GREEN}The game is over. Thank you for playing RKR on {Gamemode.CurrentGameMode}!{Colors.COLOR_RESET}");
    }

    /// <summary>
    /// True if the game is over and the kitties have won. False if they lost.
    /// </summary>
    /// <param name="win"></param>
    private static void GameStats(bool win)
    {
        var difficulty = (DifficultyLevel)Difficulty.DifficultyValue;
        foreach (var kitty in Globals.ALL_KITTIES)
        {
            if (win) StatManager.IncrementWin(kitty.Value, difficulty);
            StatManager.IncrementWinStreak(kitty.Value, win);
        }
        AwardManager.AwardGameStatRewards();
    }

    public static void NotifyEndingGame()
    {
        DiscordFrame.Initialize();
        Utility.TimedTextToAllPlayers(EndingTimer, $"{Colors.COLOR_YELLOW}The game will end in {EndingTimer} seconds.{Colors.COLOR_RESET}");
        Globals.GAME_ACTIVE = false;
        Utility.SimpleTimer(EndingTimer, EndGame);
    }
}

File: src/Events/Gameover.ts
import { Kitty } from 'src/Game/Entities/Kitty/Kitty'
import { Wolf } from 'src/Game/Entities/Wolf'
import { PodiumManager } from 'src/Game/Podium/PodiumManager'
import { PodiumUtil } from 'src/Game/Podium/PodiumUtil'
import { CurrentGameMode } from 'src/Gamemodes/CurrentGameMode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { Difficulty } from 'src/Init/Difficulty/Difficulty'
import { DifficultyLevel } from 'src/Init/Difficulty/DifficultyOption'
import { Challenges } from 'src/Rewards/Challenges/Challenges'
import { AwardManager } from 'src/Rewards/Rewards/AwardManager'
import { SaveManager } from 'src/SaveSystem2.0/SaveManager'
import { Colors } from 'src/Utility/Colors/Colors'
import { Utility } from 'src/Utility/Utility'
import { GameoverUtil } from './GameoverUtil'

export class Gameover {
    public static NoEnd: boolean = false

    public static GameOver(): boolean {
        return Gameover.WinningGame() || Gameover.LosingGameCheck()
    }

    private static WinningGame(): boolean {
        if (!Globals.WinGame) return false
        Gameover.SendWinMessage()
        Gameover.GameStats(true)
        GameoverUtil.SetColorData()
        GameoverUtil.SetBestGameStats()
        GameoverUtil.SetFriendData()
        Gameover.StandardWinChallenges()
        Gameover.SaveGame()
        print(`${Colors.COLOR_GREEN}Stay a while for the end game awards!!${Colors.COLOR_RESET}`)
        Utility.SimpleTimer(5.0, PodiumManager.BeginPodiumEvents)
        return true
    }

    private static StandardWinChallenges = () => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        Challenges.NecroWindwalk()
        Challenges.BlueFire()
        Challenges.PinkFire()
        Challenges.WhiteTendrils()
        Challenges.ZandalariKitty()
        Challenges.FreezeAura()
    }

    private static LosingGame = () => {
        Wolf.RemoveAllWolves()
        GameoverUtil.SetColorData()
        GameoverUtil.SetFriendData()
        Gameover.GameStats(false)
        Gameover.SaveGame()
        PodiumUtil.NotifyEndingGame()
    }

    private static SaveGame = () => {
        Utility.SimpleTimer(1.5, SaveManager.SaveAll)
        Utility.SimpleTimer(2.5, SaveManager.SaveAllDataToFile)
    }

    private static EndGame = () => {
        for (const player of Globals.ALL_PLAYERS) CustomVictoryBJ(player.handle, true, true)
    }

    private static LosingGameCheck(): boolean {
        if (CurrentGameMode.active !== GameMode.Standard) return false
        if (Gameover.NoEnd) return false

        for (let i = 0; i < Globals.ALL_PLAYERS.length; i++) {
            const kitty = Globals.ALL_KITTIES.get(Globals.ALL_PLAYERS[i])!
            if (kitty.isAlive()) return false
        }
        Gameover.LosingGame()
        return true
    }

    private static SendWinMessage = () => {
        if (CurrentGameMode.active === GameMode.Standard)
            print(
                `${Colors.COLOR_GREEN}Congratulations on winning the game on ${Difficulty.DifficultyOption.toString()}!${Colors.COLOR_RESET}`
            )
        else
            print(
                `${Colors.COLOR_GREEN}The game is over. Thank you for playing RKR on ${CurrentGameMode.active}!${Colors.COLOR_RESET}`
            )
    }

    /// <summary>
    /// True if the game is over and the kitties have won. False if they lost.
    /// </summary>
    /// <param name="win"></param>
    private static GameStats = (win: boolean) => {
        for (const [_, kitty] of Globals.ALL_KITTIES) {
            Gameover.IncrementGameStats(kitty)
            if (win) Gameover.IncrementWins(kitty)
            Gameover.IncrementWinStreak(kitty, win)
        }
        AwardManager.AwardGameStatRewards()
    }

    private static IncrementGameStats = (kitty: Kitty) => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        const stats = kitty.SaveData.GameStats
        switch (Difficulty.DifficultyValue) {
            case DifficultyLevel.Normal:
                stats.NormalGames += 1
                break

            case DifficultyLevel.Hard:
                stats.HardGames += 1
                break

            case DifficultyLevel.Impossible:
                stats.ImpossibleGames += 1
                break
            case DifficultyLevel.Nightmare:
                stats.NightmareGames += 1
                break
        }
    }

    private static IncrementWins = (kitty: Kitty) => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        const stats = kitty.SaveData.GameStats
        switch (Difficulty.DifficultyValue) {
            case DifficultyLevel.Normal:
                stats.NormalWins += 1
                break

            case DifficultyLevel.Hard:
                stats.HardWins += 1
                break

            case DifficultyLevel.Impossible:
                stats.ImpossibleWins += 1
                break
            case DifficultyLevel.Nightmare:
                stats.NightmareWins += 1
                break
        }
    }

    private static IncrementWinStreak = (kitty: Kitty, win: boolean) => {
        if (CurrentGameMode.active !== GameMode.Standard) return
        const stats = kitty.SaveData.GameStats

        if (win) {
            stats.WinStreak += 1
            if (stats.WinStreak > stats.HighestWinStreak) stats.HighestWinStreak = stats.WinStreak
        } else stats.WinStreak = 0
    }
}
